chainstore.go

```diff
@@ -1,147 +1,231 @@
 package chainstore
 
 import (
-	"errors"
-	"io/ioutil"
+	"fmt"
 	"regexp"
+	"sync"
+
+	"golang.org/x/net/context"
 )
 
-var (
-	ErrInvalidKey = errors.New("Invalid key")
+type storeFn func(s Store) error
 
-	KeyInvalidator = regexp.MustCompile(`(i?)[^a-z0-9\/_\-:\.]`)
+var (
+	keyInvalidator = regexp.MustCompile(`(i?)[^a-z0-9\/_\-:\.]`)
 )
 
 const (
-	MaxKeyLen = 256
+	maxKeyLen = 256
 )
 
 type Store interface {
 	Open() error
 	Close() error
-	Put(key string, val []byte) error
-	Get(key string) ([]byte, error)
-	Del(key string) error
+	Put(ctx context.Context, key string, val []byte) error
+	Get(ctx context.Context, key string) ([]byte, error)
+	Del(ctx context.Context, key string) error
 }
 
-// TODO: how can we check if a store has been opened...?
+type storeWrapper struct {
+	Store
+	errE  error
+	errMu sync.RWMutex
```

Comment:
Why do we need this mutex? Is it even possible that anything would be writing error to one store at the same time?